### DISCUSSION TITLE
Static site generator tech stack survey for the primary website

### DISCUSSION THREAD

**Discussion by GarthDB** [DesignSystemsEngineer/designsystemsengineer.github.io#36] [Polls]
To keep things simple and free, this main website should be planned as a static site with collections of Markdown files. This would make it easier for others to contribute ideas, articles, etc. We could switch in the future if needed, but I am curious about opinions on starting a tech stack.

1. [Jekyll](https://jekyllrb.com/)
  * Built-in support in GH Pages (no build step)
  * Ruby
  * Some allowed [plugins](https://docs.github.com/en/pages/setting-up-a-github-pages-site-with-jekyll/about-github-pages-and-jekyll#plugins)
2. [Eleventy](https://www.11ty.dev/)
  * [Performant builds](https://www.11ty.dev/docs/performance/)
  * Pretty popular with good resources
3. [Gatsby](https://www.gatsbyjs.com/)
4. [Next.js](https://nextjs.org/)
5. [Astro](https://astro.build/)

Feel free to comment below with additional recommendations or reasons for picking one over another. This poll won't be the definitive decision maker because it's ultimately most swayed by those who will be maintaining the website.

**Comment by ddamato** [comment]
I am a huge Astro fanboy; I've been using it for all my new sites and slowly migrating older sites to use it. It's probably more fully featured than most of the list, but not overkill like Next.js.

**Reply by resource11** [reply]
Seconding Astro here. I’ve migrated from Gatsby to Astro and the depth of features and developer experience Astro provides will set up a nice scalable solution for y’all.

**Reply by mattfelten** [reply]
3rding Astro. It's very good

**Reply by basher** [reply]
4thing Astro.

**Reply by jnjosh** [reply]
All this Astro excitement is now just making me want to migrate my Hugo theme to Astro to see what everyone is excited about!

**Reply by resource11** [reply]
Content collections, dynamic route generation, support for type-safe MDX... and more!

**Comment by zastrow** [comment]
Gotta put my vote for Eleventy. It is my favorite SSG that I have ever used and it is fast. Plus, the community around it is one of the best: friendly, curious, and smart!

**Comment by Westbrook** [comment]
Useful prior art in a study like this https://www.zachleat.com/web/site-generator-review/

**Reply by GarthDB** [reply]
@Westbrook, thanks for the resource. Do you have any feelings about what it says? I mean, I definitely read all of the words, but let's pretend someone else didn't and likes hearing your opinion, hypothetically. What would you use?

**Reply by zastrow** [reply]
@GarthDB Best thing I have read all day 🤣

**Reply by Westbrook** [reply]
Haha. I voted in the survey…

I use and would likely use again Eleventy for this sort of effort. You didn’t list it, but I might put a little time into Enhance before finalizing that decision. Both have pretty similar end goals but some of the thing in the middle dynamism, management, etc could blow the ship one way or another. 

Most of my decisions would start with the lowest amount of default KBs for the page request, which would dump Next and Gatsby to get started. Beyond that you get into nuances of how you want to manage your content, when/how often you want to publish/update it, the dynamism of your content over time. The specific goals of which may move you back and forth between the faster/lighter options.

**Reply by GarthDB** [reply]
@Westbrook  I would love to use [Enhance](https://enhance.dev), but does it support static sites? I was planning to host on gh-pages.

**Reply by Westbrook** [reply]
Indeed. It does need some sort of server/serverless more than just gh-pages. So, back to my original answer!

**Reply by GarthDB** [reply]
@kristoferjoseph I think this is a request for adding support for static site builds to Enhance. Who else should I tag in this? Brian? Simon?

**Reply by GarthDB** [reply]
@Westbrook, @kristoferjoseph just sent me this [link](https://github.com/ryanbethel/ssg-enhance); a proof of concept for Enhance SSG.

**Comment by o-t-w** [comment]
I would choose between Eleventy and Astro. 

Jekyll is written in Ruby so there’s less chance of outside contributions than a tool that uses JavaScript, a language which far more people know. 

Gatsby is imo the poster child of the very worst excesses of hype-driven development and the over-complication of frontend. 

There’s a lot of dynamism within Astro and it has a funded full-time team of top-tier developers. 

Eleventy is the simplest solution around. It makes fewer breaking changes than Astro.

**Reply by GarthDB** [reply]
Jekyll seems looked down upon by the wider front-end community because of Ruby, but I feel like Jekyll's built-in nature is overlooked as a nice benefit. I'm not saying it's enough to sway things completely, but I've always appreciated that aspect.

**Comment by jnjosh** [comment]
I voted for Eleventy but just want to call out how awesome [Hugo](https://gohugo.io/) is. I do all my sites on Hugo and it's super fast! From the Eleventy build performance page:

<img width="858" alt="Screenshot 2024-05-01 at 2 25 13 PM" src="https://github.com/DesignSystemsEngineer/designsystemsengineer.github.io/assets/66142/43cae7b5-71d3-4c8e-bf42-466d1e0c9485">

**Reply by ddamato** [reply]
I'm curious how much we should lean into build performance as a metric to track versus site performance? I think I'd argue that between the options, the user experience should outweigh the developer experience (but yes, why not both).

**Reply by GarthDB** [reply]
@ddamato I agree that the user's experience should be pretty high on the priority list.

I would add that these are all static site generators so hopefully we can keep things pretty performant for the end user.

Also, I'm planning to have it built and deployed using GitHub actions so many content authors won't need to worry about build performance. It would most impact the front end developers working on it locally.

**Reply by jnjosh** [reply]
That's fair—in my experience running a site like this in the past, fast iteration/publishing was helpful, but not everything. Especially when that difference is all under 1 minute. Hugo leans more into blog structure often too so I'm not going to argue for it beyond mentioning it. 

Anyway, like I said, I voted for Eleventy which seems pretty good as far as I've seen. :)
